pymc/distributions/distribution.py

```diff
@@ -277,6 +314,20 @@ def __new__(
                 "for a standalone distribution."
             )
 
+        if "name" in kwargs:
+            name = kwargs.pop("name")
+        elif len(args) > 0 and isinstance(args[0], string_types):
+            name = args[0]
+            args = args[1:]
+        else:
+            name = _extract_target_of_assignment(2)
```

Comment:
For situations where automated name determination fails, this part should raise an informative error.
For example, one might simply forget to pass the name:

```python
with pmodel:
    pm.Normal("Y", pm.Uniform(), observed=2)
```

Reply:
or even better: make `_extract_target_of_assignment` raise the informative error and unit test that directly.

Reply:
Good idea. What do you mean by an informative error message? I'm not sure what make can I say beyond that I failed to infer the name, and in the `_extract_target_of_assignment` that no assignment instruction was found at that frame depth.

Reply:
Maybe something about "inferring the name from the outer code"

if you have a handle on the line of code print that

definitely instructions what do do about it: please pass a name as the first argument

Reply:
Just pushed. Does that look ok?

Reply:
the message looks good, but I'd recommend to raise it in the extraction function instead of returning `None`. It gives a cleaner signature (str or raise) and separation of responsibility..

Reply:
I think the error message changes. Are we ok with an Assignment not found error that is raised, caught, and re-raised as a Name needs to be provided error message?